### ISSUE TITLE
TensorFlow 2 slow on small numbers of batches

### ISSUE THREAD

**Issue by seermer** [tensorflow/tensorflow#48844]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10 and Colab (observed such performance issue on both platform)
- TensorFlow installed from (source or binary): pip / provided in colab
- TensorFlow version (use command below): 2.4.1
- Python version: 3.8
- CUDA/cuDNN version: CUDA 11.0, cuDNN 8.0 / provided in colab
- GPU model and memory: GTX 1660Ti / provided in colab

I have already asked this question on StackOverflow, but nobody answered, so considering it is a performance issue, I opened an issue, sorry if this is not appropriate to be asked in issues.

When training a small number of batches (not batch size), TensorFlow 2 is significantly slower than PyTorch.
(for example, train 30 batches, note that different but still relatively large ratios still apply for a smaller number of batches like training a single batch, different models with a few more layers/params, and/or different batch size)
below are approximations of the performance difference I observed in colab CPU:

with low-level gradient tape training, TensorFlow is 9 times slower than equivalent model/data on PyTorch
with low-level gradient tape + @tf.function train step, TensorFlow is 18 times slower than equivalent model/data on PyTorch
with high-level fit(), TensorFlow is 19 times slower than equivalent model/data on PyTorch
with high-level train_on_batch(), TensorFlow is 22 times slower than equivalent model/data on PyTorch

**A reproducible code** example can be found here, which shows a simple model (can run on both CPU or GPU depending on whether GPU is turned on in Colab) and generated random data being trained with all functions/API discussed above: https://colab.research.google.com/drive/1VHYLEZ79etKSkK0iWIqHkt7OG8nbU9uU?usp=sharing 

above code is partially modified from the official tutorial https://www.tensorflow.org/guide/keras/writing_a_training_loop_from_scratch
and partially written on my own.

training single batch or a few batches are widely used in Deep Reinforcement Learning and some special cases of Deep Learning. 
The code above is just for demonstration of my observation. I also noticed a large train time gap between TF and PyTorch when I work on DQN locally on my Windows machine, since DQN Experience replay trains on one batch each time.

Am I using TF incorrectly? if yes, what is the best way to train a small number of batches?
Even if I am not using the best way. 
Why is @tf.function that is meant to accelerate slowing down training?
Why is train_on_batch that seems to be designed for a single batch the slowest of all?

Any help would be great! Thank you.

**Comment by tilakrayal** [comment]
@jvishnuvardhan ,

I was able to reproduce the code.Please find the [gist](https://colab.research.google.com/gist/tilakrayal/74e54faafe4e743a8f2b4538481c6db9/48844.ipynb) here

**Comment by mdanatg** [comment]
Your findings are mostly expected, I will include some explanations below. At a glance, I don't see any obvious issues with the way you've been constructing the code.

1. TF eager

TF eager is know to be slower compared to PyTorch. Your measurement of about an order of magnitude seems in the ballpark. The rule of thumb is to use eager execution for debugging, and production code to use tf.function (see below for its performance considerations)

2. Keras

I'm actually not sure why the Keras versions which use model.fit and model.train_on_batch are not faster, even with a warm-up call. Internally, they should use `tf.function`, but then again, they may perform additional work (like summaries or whatnot).
cc @tomerk @fchollet for additional insight

3. tf.function

The most important thing to know about tf.function is that it compiles just-in-time, that is, it's actually compiled the first time you call it (traced in TF parlance, see [here](https://www.tensorflow.org/guide/intro_to_graphs) and [here](https://www.tensorflow.org/guide/function) for some more high-level info). Tracing is considerably slow compared to executing. However, once traced, subsequent executions will be much faster.

Long story short: the initial call of tf.function is usually much slower, as your test revealed. So if you have a workflow that repeatedly builds a function and calls it only a small number of times, it will not be faster than eager. Usually, you'd notice this in unit tests. But if you have a way to reuse the same function for multiple runs, then it should quickly pay off.

Now, if we modify our benchmark to account for that by running a warm-up call:

```
train_step(inputs, labels)  # Do a warm-up run for tracing

t = time.time()
for _ in range(30):
  loss_v = train_step(inputs, labels)
print(time.time() - t)
```

Then we get speed roughly on par with PyTorch. It's slightly faster, but in the same order of magnitude.

So the main caveat is the function tracing, which should be a one-time cost.

More on `tf.function`: to get the state-of-the art speed, you should consider enabling XLA compilation, by specifying `@tf.function(experimental_compile=True)` or `@tf.function(jit_compile=True)` in newer versions. That is restricted to a subset of TensorFlow (it works in our case), but gives you the best available performance.

Even more on `tf.function`: a good way to get even greater performance is to run several training steps in the same graph:

```
@tf.function
def train_loop(n_steps):
  for _ in tf.range(n_steps):
    loss_v = train_step(inputs, labels)
  return loss_v

train_loop(tf.constant(1))  # Again, a warm-up for tracing

# Note: use tf.constant(1) and tf.constant(30), not 1 and 30, to avoid retracing

t = time.time()
train_loop(tf.constant(30))
print(time.time() - t)
```

This gave the best speed-up in my tests.
